File: python/string.py
def z_func(s: str) -> list:
    n = len(s)
    z = [0]*(n)
    l, r = 0, 0
    for i in range(1, n):
        if i < r:
            z[i] = min(r-i, z[i-l])
        while i+z[i] < n and s[z[i]] == s[i+z[i]]:
            z[i] += 1
        if i+z[i] > r:
            l, r = i, i+z[i]
    return z


def find_position(t: str, p: str) -> list:
    """
    t:text string,p:pattern string
    from t find p
    """
    idx = []
    Lt, Lp = len(t), len(p)
    s = p+"#"+t
    z = z_func(s)
    for i in range(Lt):
        k = z[i+Lp+1]
        if k == Lp:
            idx.append(i)
    return idx


def prefix_func(s: str) -> list:
    n = len(s)
    pi = [0]*n
    for i in range(1, n):
        j = pi[i-1]
        while j > 0 and s[i] != s[j]:
            j = pi[j-1]
        if s[i] == s[j]:
            j += 1
        pi[i] = j
    return pi


def match(t: str, p: str) -> bool:
    """
    t:text string,p:pattern string
    from t find p
    """
    lps = prefix_func(p+'#'+t)
    Lt, Lp = len(t), len(p)
    for i in range(Lp+1, Lt+Lp+1):
        if lps[i] == Lp:
            return True
    return False

File: cpp/string.cpp
#include <vector>
#include <string>
namespace internal
{
    std::vector<int> z_func(const std::string &s)
    {
        int n = s.size();
        std::vector<int> z(n);
        int l = 0, r = 0;
        for (int i = 1; i < n; i++)
        {
            if (i < r)
            {
                z[i] = std::min(r - i, z[i - l]);
            }
            while (i + z[i] < n && s[z[i]] == s[i + z[i]])
            {
                z[i]++;
            }
            if (i + z[i] > r)
            {
                l = i;
                r = i + z[i];
            }
        }
        return z;
    }
    std::vector<int> find_occur(const std::string &t, const std::string &p)
    {
        std::vector<int> idx;
        int len_p = p.size(), len_t = t.size();
        std::string s = p + "#" + t;
        auto z = z_func(s);
        for (int i = 0; i < len_t; i++)
        {
            int k = z[i + len_p + 1];
            if (k == len_p)
            {
                idx.emplace_back(i);
            }
        }
        return idx;
    }

    std::vector<int> prefix_func(const std::string &s)
    {
        int n = s.size();
        std::vector<int> pi(n);
        for (int i = 1; i < n; i++)
        {
            int j = pi[i - 1];
            while (j > 0 && s[i] != s[j])
            {
                j = pi[j - 1];
            }
            if (s[i] == s[j])
            {
                j++;
            }
            pi[i] = j;
        }
        return pi;
    }

    bool match(const std::string &t, const std::string &p)
    {
        auto lps = prefix_func(p + "#" + t);
        int sz1 = t.size(), sz2 = p.size();
        for (int i = sz2 + 1; i <= sz1 + sz2; i++)
        {
            if (lps[i] == sz2)
            {
                return true;
            }
        }
        return false;
    }
}